### ISSUE TITLE
drawViewHierarchyInRect:afterScreenUpdates: & Autolayout

### ISSUE THREAD

**Issue by tombatesuk** [facebook/ios-snapshot-test-case#66]
It would seem that we have been relying on the drawViewHierarchyInRect:afterScreenUpdates: that was called before being removed in #63.
The specific issue is that a view is marked and not installed in the storyboard for some size classes but is still appearing is now appearing.
The work around we are currently using through lack of a better alternative is to just call:
```
[subject.view drawViewHierarchyInRect:subject.view.frame afterScreenUpdates:TRUE];
```
before our expectation.
This is obviously not ideal. Does anyone know what this is actually doing under the hood or have any better suggestions to fix this issue?

**Comment by dstnbrkr** [comment]
Hi @tombatesuk - it seems drawViewHierarchyInRect:afterScreenUpdates: has side-effects that some tests depend on and others fail as a result of. I'm curious to find out what's going on under the hood so we can apply those side effects directly or at least better understand them.

If you could provide a small failing test case (or one that only succeeds when you call drawViewHierarchyInRect:afterScreenUpdates: before verifying the view), that would definitely help pinpoint which side effect storyboards and autolayout require.

**Comment by tombatesuk** [comment]
@dstnbrkr I will look at creating and small test project with a view and test that mimics the results that we are experiencing and send it over.

**Comment by tombatesuk** [comment]
@dstnbrkr Here is a test project that shows how using drawViewHierarchyInRect:afterScreenUpdates: can affect the snapshot.
The project has two reference images for the same view controller. In this instance the spec using drawViewHierarchyInRect:afterScreenUpdates: is the correct snapshot.
https://github.com/tombatesuk/snapshot-test

**Comment by dstnbrkr** [comment]
Thanks @tombatesuk, I see the issue. Because you're testing a UIView via a UIViewController via a UIStoryboard, the view you're expecting to see is the state of the view one runloop tick later.

Here's what's happening in the passing test:
- instantiate controller
- drawViewHierarchyInRect:afterUpdates:YES is called // blocks for one run loop tick
- snapshot test passes (extra runloop tick waited for constraints and layout to be applied)

and in the failing test:
- instantiate controller
- snapshot test fails (nobody told the view how to layout yet)

Basically this test depends on async behavior and the test needs to wait one runloop tick for the view to get into the state you want to test. I don't believe drawViewHierarchyInRect:afterUpdates:YES is mutating your view in any way, it's the afterUpdates:YES that's making the test pass by forcing the test to wait one more runloop tick. You wouldn't want to do this in every test, only tests where the view under test is async somehow.

**Comment by dstnbrkr** [comment]
The tl;dr on the above is:

It's correct for tests to use drawViewHierarchyInRect:afterUpdates:YES or any other means of ticking the runloop when the view under test is async.
